### PR TITLE
Write a minimal MIR opt chapter

### PR DESCRIPTION
**Pull request by mark-i-m**
cc @oli-obk @JohnTitor #331 

This is a really minimal version until we have something more later.

### CODE REVIEW

File: src/mir/optimizations.md [hunk 1]
```diff
@@ -1 +1,45 @@
 # MIR optimizations
+
+MIR optimizations are optimizations run on the [MIR][mir] to produce better MIR
+before codegen. This is important for two reasons: first, it make the final
+generated executable code better, and second, it means that LLVM has less work
+to do, so compilation is faster.
```
Comment (oli-obk) [inline src/mir/optimizations.md]: This part is particularly effective, because we can optimize generic functions, so all the monomorphisations are cheaper

Comment (mark-i-m) [reply src/mir/optimizations.md]: Oh, that's cool. I added a note.

File: src/mir/optimizations.md [hunk 2]
```diff
@@ -1 +1,49 @@
 # MIR optimizations
+
+MIR optimizations are optimizations run on the [MIR][mir] to produce better MIR
+before codegen. This is important for two reasons: first, it make the final
```
Comment (JohnTitor) [inline src/mir/optimizations.md]:
```suggestion
before codegen. This is important for two reasons: first, it makes the final
```

File: src/mir/optimizations.md [hunk 3]
```diff
@@ -1 +1,49 @@
...
+[monomorph]: https://rust-lang.github.io/rustc-guide/appendix/glossary.html?highlight=monomorphize#appendix-c-glossary
+
+MIR optimizations run after borrow checking. We run a series of optimization
+passes over the MIR to improve it. Some passes are required to run on all code,
+some passes don't actually do optimizations but only check stuff, and some
+passes are only turned on in `release` mode.
+
+The [`optimized_mir`][optmir] [query] is called to produce the optimized MIR
+for a given [`DefId`][defid]. This query makes sure that the borrow checker has
+run and that some validation has occured. Then, it [steals][steal] the MIR
```
Comment (JohnTitor) [inline src/mir/optimizations.md]:
```suggestion
run and that some validation has occurred. Then, it [steals][steal] the MIR
```

**Comment by mark-i-m** [comment]
Could someone please approve this PR so I can merge it? Thanks

File: src/mir/optimizations.md [hunk 4]
```diff
@@ -1 +1,49 @@
...
+run.  Each pass in the array is a struct that implements the [`MirPass`] trait.
+The array is an array of `&dyn MirPass` trait objects. Typically, a pass is
+implemented in its own submodule of the [`rustc_mir::transform`][trans] module.
+
+[rop]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/transform/fn.run_optimization_passes.html
+[`MirPass`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/transform/trait.MirPass.html
+[trans]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/transform/index.html 
+
+Some examples of passes are:
+- `CleanupNonCodegenStatements`: remove some of the info that is only need for
```
Comment (spastorino) [inline src/mir/optimizations.md]:
```suggestion
- `CleanupNonCodegenStatements`: remove some of the info that is only needed for
```

**Comment by mark-i-m** [comment]
@spastorino Sorry! I forgot about these comments. You are right on both of them. Could you open a PR please?

**Comment by spastorino** [comment]
@mark-i-m done https://github.com/rust-lang/rustc-guide/pull/520

**Pull request merged**